### ISSUE TITLE
device offline whereas device seems ready

### ISSUE THREAD

**Issue by gigaga** [bitrise-steplib/steps-wait-for-android-emulator#13]
Hi,

I have a problem with wait-for-android-emulator step. Currently, the step seems indicates that device is offline whereas, `adb devices` indicates that is ready. Also, the build stays blocked... :(
To reproduce, I created 3 steps:

- avd-manager@0.9.2 
- script@1.1.5: contains just `adb devices` command
- wait-for-android-emulator@1.0.4

Related build: https://app.bitrise.io/build/b4dbecc4f793ab57

The complete logs [b4dbecc4f793ab57.log](https://github.com/bitrise-steplib/steps-wait-for-android-emulator/files/2607862/b4dbecc4f793ab57.log)

Best regards,

**Comment by fehersanyi-bitrise** [comment]
Hello there! @gigaga 

The avd manager is not really reliable. I would recommend testing with the `Virtual device testing` step, with a guide here to help you set it up: https://devcenter.bitrise.io/testing/device-testing-for-android/

**Comment by fehersanyi-bitrise** [comment]
@gigaga You are right, I remember. I will ask the tooling team to pick it up and do something about this .

**Comment by gigaga** [comment]
Also, I am going to try to find a workaround... 😉

**Comment by gigaga** [comment]
I found a workaround by implementing a script step with the following content:

```
#!/usr/bin/env bash
# fail if any commands fails
set -e
# debug log
#set -x

cd $ANDROID_HOME/emulator

echo no | avdmanager create avd -n Nexus_5X_API_26 -k "system-images;android-26;google_apis;x86" --force

emulator -avd Nexus_5X_API_26 -no-window -no-audio -debug-init > /dev/null 2>&1 &

echo "Waiting emulator is ready..."

adb wait-for-device shell 'while [[ -z $(getprop sys.boot_completed) ]]; do sleep 1; done; input keyevent 82'

echo "Emulator is ready!"
```

**Comment by viktorbenei** [comment]
Thanks for sharing the workaround @gigaga !

**Comment by gigaga** [comment]
@viktorbenei you're welcome 😁

**Comment by gigaga** [comment]
Hi @TannerJuby1,
I'm using this script in replacement of those steps 😉

**Comment by TannerJuby1** [comment]
@gigaga thanks! I'm attempting to do this but it's been stalled on `waiting emulator is ready...`. The loading of remote repository is getting stalled on 25%:

```
Loading local repository...                                                     
[=========                              ] 25% Loading local repository...       
[=========                              ] 25% Fetch remote repository...        
[=========                              ] 25% Fetch remote repository...        
[=========                              ] 25% Fetch remote repository...        
[=======================================] 100% Fetch remote repository...   
``` 

And has been like this for about 15 minutes... Is there any addition set up you have to do within Bitrise, or is this suppose to work like a plug and play?

**Comment by gigaga** [comment]
Can you provide me the full related logs?
I think that the emulator is not started.

**Comment by TannerJuby1** [comment]
```
Agent version:	0.6.3
Start time:	2019-01-15T18:30:48Z
$ docker "run" "-v" "/var/run/docker.sock:/var/run/docker.sock" "-v" "/bitrise:/bitrise" "-v" "/tmp/bitrise-agent898411941:/.bitrise-config" "--privileged" "--network=host" "--name=bitrise-main-container" "bitriseio/android-ndk:pinned" "bitrise" "run" "--config" "/.bitrise-config/bitrise.yml" "--inventory" "/.bitrise-config/bitrise.secrets.yml" "--json-params-base64" "eyJ3b3JrZmxvdyI6InJ1bl90ZXN0cyJ9"

  ██████╗ ██╗████████╗██████╗ ██╗███████╗███████╗
  ██╔══██╗██║╚══██╔══╝██╔══██╗██║██╔════╝██╔════╝
  ██████╔╝██║   ██║   ██████╔╝██║███████╗█████╗
  ██╔══██╗██║   ██║   ██╔══██╗██║╚════██║██╔══╝
  ██████╔╝██║   ██║   ██║  ██║██║███████║███████╗
  ╚═════╝ ╚═╝   ╚═╝   ╚═╝  ╚═╝╚═╝╚══════╝╚══════╝

[32;1m  version: 1.26.0[0m

[36mINFO[0m[18:30:52] [33;1mbitrise runs in Secret Filtering mode[0m 
[36mINFO[0m[18:30:52] [33;1mbitrise runs in CI mode[0m           
[36mINFO[0m[18:30:52] Running workflow: [32;1mrun_tests[0m       

[34;1mSwitching to workflow:[0m run_tests

+------------------------------------------------------------------------------+
| (0) avd-manager@0.9.2                                                        |
+------------------------------------------------------------------------------+
| id: avd-manager                                                              |
| version: 0.9.2                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-01-15T18:30:56Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
[36mINFO[0m[18:30:56] Start installing (golang) with apt-get       
[36mINFO[0m[18:30:56]  * [32;1m[OK][0m Step dependency (go) installed, available. 

[34;1mConfigs:[0m
- Version: 26
- Resolution: 240x426
- Density: 120
- Orientation: portrait
- Tag: google_apis
- ABI: x86
- Profile: pixel_xl
- ID: emulator
- CustomCommandFlags: -no-window -no-audio -debug-init
- Overwrite: true
- CustomConfig:
disk.dataPartition.size=1024M
hw.camera.back=emulated
hw.camera.front=emulated
hw.keyboard=yes
hw.gpu.enabled=true
hw.gpu.mode=swiftshader
hw.ramSize=1536
runtime.network.latency=none
runtime.network.speed=full
showDeviceFrame=no
skin.dynamic=yes
skin.path=_no_skin
skin.path.backup=_no_skin
vm.heapSize=384

[34;1mEnsure sdk packages: [tools platform-tools system-images;android-26;google_apis;x86][0m
[32;1m- Done[0m

[34;1mCreate AVD[0m
[32;1m- Done[0m

[34;1mStart emulator[0m
- Device with serial: emulator-5554 started
[32;1m- Done[0m
|                                                                              |
+---+---------------------------------------------------------------+----------+
| [32;1m✓[0m | [32;1mavd-manager@0.9.2[0m                                             | 145 sec  |
+---+---------------------------------------------------------------+----------+

                                          ▼

+------------------------------------------------------------------------------+
| (1) activate-ssh-key@4.0.3                                                   |
+------------------------------------------------------------------------------+
| id: activate-ssh-key                                                         |
| version: 4.0.3                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-01-15T18:33:22Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
[36mINFO[0m[18:33:22] Start installing (expect) with apt-get       
[36mINFO[0m[18:33:22]  * [32;1m[OK][0m Step dependency (expect) installed, available. 
[36mINFO[0m[18:33:22] Start installing (git) with apt-get          
[36mINFO[0m[18:33:22]  * [32;1m[OK][0m Step dependency (git) installed, available. 
[34;1mConfig:[0m
- SSHRsaPrivateKey: *****
- SSHKeySavePath: /root/.ssh/bitrise_step_activate_ssh_key
- IsRemoveOtherIdentities: true
- Verbose: false

$ ssh-add "-l"
Could not open a connection to your authentication agent.
ssh_agent_check_result: 2
ssh-agent not started
starting a new ssh-agent and exporting connection information with envman
$ ssh-agent
Expose SSH_AUTH_SOCK for the new ssh-agent, with envman
$ bash "-c" "/tmp/spawn475262101/tmp_spawn.sh"
spawn ssh-add /root/.ssh/bitrise_step_activate_ssh_key
Identity added: /root/.ssh/bitrise_step_activate_ssh_key (/root/.ssh/bitrise_step_activate_ssh_key)

[32;1mSuccess[0m
The SSH key was saved to /root/.ssh/bitrise_step_activate_ssh_key
and was successfully added to ssh-agent.
|                                                                              |
+---+---------------------------------------------------------------+----------+
| [32;1m✓[0m | [32;1mactivate-ssh-key@4.0.3[0m                                        | 11 sec   |
+---+---------------------------------------------------------------+----------+

                                          ▼

+------------------------------------------------------------------------------+
| (2) git-clone@4.0.14                                                         |
+------------------------------------------------------------------------------+
| id: git-clone                                                                |
| version: 4.0.14                                                              |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-01-15T18:33:32Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
[34;1mconfig:
[0m
- RepositoryURL: git@bitbucket.org:devetry/ytp_android.git
- CloneIntoDir: /bitrise/src
- Commit: 
- Tag: 
- Branch: unittests
- BranchDest: 
- PRID: 0
- PRRepositoryURL: 
- PRMergeBranch: 
- ResetRepository: false
- CloneDepth: 0
- BuildURL: https://app.bitrise.io/build/cdae9f32a35093b5
- BuildAPIToken: [REDACTED]
- UpdateSubmodules: true
- ManualMerge: true
[34;1mgit "init"[0m
Initialized empty Git repository in /bitrise/src/.git/
[34;1mgit "remote" "add" "origin" "git@bitbucket.org:devetry/ytp_android.git"[0m
[34;1mgit "fetch" "origin" "unittests"[0m
Warning: Permanently added 'bitbucket.org,18.205.93.1' (RSA) to the list of known hosts.
From bitbucket.org:devetry/ytp_android
 * branch            unittests  -> FETCH_HEAD
 * [new branch]      unittests  -> origin/unittests
[34;1mgit "checkout" "unittests"[0m
Switched to a new branch 'unittests'
Branch unittests set up to track remote branch unittests from origin.
[34;1mgit "merge" "origin/unittests"[0m
Already up-to-date.
[34;1mgit "submodule" "update" "--init" "--recursive"[0m
[34;1m
Exporting git logs
[0m
=> GIT_CLONE_COMMIT_COMMITER_EMAIL
   value: tannerjuby@Tanner-Devetry-Mac.local

=> GIT_CLONE_COMMIT_HASH
   value: e9170d3e4332da0c7687e93a9701bd62e9eedc9f

=> GIT_CLONE_COMMIT_MESSAGE_SUBJECT
   value: Removed unwanted tests

=> GIT_CLONE_COMMIT_MESSAGE_BODY
   value: 

=> GIT_CLONE_COMMIT_AUTHOR_NAME
   value: Tanner Juby

=> GIT_CLONE_COMMIT_AUTHOR_EMAIL
   value: tannerjuby@Tanner-Devetry-Mac.local

=> GIT_CLONE_COMMIT_COMMITER_NAME
   value: Tanner Juby

=> GIT_CLONE_COMMIT_COUNT
   value: 195

[32;1m
Success[0m
|                                                                              |
+---+---------------------------------------------------------------+----------+
| [32;1m✓[0m | [32;1mgit-clone@4.0.14[0m                                              | 20 sec   |
+---+---------------------------------------------------------------+----------+

                                          ▼

+------------------------------------------------------------------------------+
| (3) cache-pull@2.0.1                                                         |
+------------------------------------------------------------------------------+
| id: cache-pull                                                               |
| version: 2.0.1                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-01-15T18:33:54Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
[36mINFO[0m[18:33:54] Start installing (rsync) with apt-get        
[36mINFO[0m[18:33:54]  * [32;1m[OK][0m Step dependency (rsync) installed, available. 
[36mINFO[0m[18:33:54] Start installing (tar) with apt-get          
[36mINFO[0m[18:33:54]  * [32;1m[OK][0m Step dependency (tar) installed, available. 
2019/01/15 18:34:05 Cache pull...
2019/01/15 18:34:05 => Downloading and extracting cache archive ...
2019/01/15 18:35:15 => [DONE]
2019/01/15 18:35:15 => Took: 1m9.42381905s
2019/01/15 18:35:15 => Finished
|                                                                              |
+---+---------------------------------------------------------------+----------+
| [32;1m✓[0m | [32;1mcache-pull@2.0.1[0m                                              | 87 sec   |
+---+---------------------------------------------------------------+----------+

                                          ▼

+------------------------------------------------------------------------------+
| (4) install-missing-android-tools@2.3.4                                      |
+------------------------------------------------------------------------------+
| id: install-missing-android-tools                                            |
| version: 2.3.4                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-01-15T18:35:17Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
[36mINFO[0m[18:35:17] Start installing (golang) with apt-get       
[36mINFO[0m[18:35:17]  * [32;1m[OK][0m Step dependency (go) installed, available. 

[34;1mConfigs:[0m
- GradlewPath: ./gradlew
- AndroidHome: /opt/android-sdk-linux
- NDKRevision: 

[34;1mPreparation[0m
Set executable permission for gradlew

[34;1mClearing NDK environment[0m
Unset ANDROID_NDK_HOME
Initialize Android SDK
Ensure android licences

[34;1mEnsure required Android SDK components[0m

[32;1mRequired SDK components are installed[0m
|                                                                              |
+---+---------------------------------------------------------------+----------+
| [32;1m✓[0m | [32;1minstall-missing-android-tools@2.3.4[0m                           | 45 sec   |
+---+---------------------------------------------------------------+----------+

                                          ▼

+------------------------------------------------------------------------------+
| (5) script@1.1.5                                                             |
+------------------------------------------------------------------------------+
| id: script                                                                   |
| version: 1.1.5                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: bash                                                                |
| time: 2019-01-15T18:36:02Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
Loading local repository...                                                     
[=========                              ] 25% Loading local repository...       
[=========                              ] 25% Fetch remote repository...        
[=========                              ] 25% Fetch remote repository...        
[=========                              ] 25% Fetch remote repository...        
[=======================================] 100% Fetch remote repository...       
Parsing /opt/android-sdk-linux/add-ons/addon-google_apis-google-21/package.xmlParsing /opt/android-sdk-linux/add-ons/addon-google_apis-google-22/package.xmlParsing /opt/android-sdk-linux/add-ons/addon-google_apis-google-23/package.xmlParsing /opt/android-sdk-linux/build-tools/17.0.0/package.xmlParsing /opt/android-sdk-linux/build-tools/19.1.0/package.xmlParsing /opt/android-sdk-linux/build-tools/21.1.2/package.xmlParsing /opt/android-sdk-linux/build-tools/22.0.1/package.xmlParsing /opt/android-sdk-linux/build-tools/23.0.3/package.xmlParsing /opt/android-sdk-linux/build-tools/24.0.3/package.xmlParsing /opt/android-sdk-linux/build-tools/25.0.3/package.xmlParsing /opt/android-sdk-linux/build-tools/26.0.1/package.xmlParsing /opt/android-sdk-linux/build-tools/26.0.2/package.xmlParsing /opt/android-sdk-linux/build-tools/27.0.0/package.xmlParsing /opt/android-sdk-linux/build-tools/27.0.1/package.xmlParsing /opt/android-sdk-linux/build-tools/27.0.2/package.xmlParsing /opt/android-sdk-linux/build-tools/27.0.3/package.xmlParsing /opt/android-sdk-linux/build-tools/28.0.0/package.xmlParsing /opt/android-sdk-linux/build-tools/28.0.1/package.xmlParsing /opt/android-sdk-linux/build-tools/28.0.2/package.xmlParsing /opt/android-sdk-linux/build-tools/28.0.3/package.xmlParsing /opt/android-sdk-linux/emulator/package.xmlParsing /opt/android-sdk-linux/extras/android/m2repository/package.xmlParsing /opt/android-sdk-linux/extras/google/google_play_services/package.xmlParsing /opt/android-sdk-linux/extras/google/m2repository/package.xmlParsing /opt/android-sdk-linux/extras/m2repository/com/android/support/constraint/constraint-layout-solver/1.0.1/package.xmlParsing /opt/android-sdk-linux/extras/m2repository/com/android/support/constraint/constraint-layout-solver/1.0.2/package.xmlParsing /opt/android-sdk-linux/extras/m2repository/com/android/support/constraint/constraint-layout/1.0.1/package.xmlParsing /opt/android-sdk-linux/extras/m2repository/com/android/support/constraint/constraint-layout/1.0.2/package.xmlParsing /opt/android-sdk-linux/patcher/v4/package.xmlParsing /opt/android-sdk-linux/platform-tools/package.xmlParsing /opt/android-sdk-linux/platforms/android-15/package.xmlParsing /opt/android-sdk-linux/platforms/android-17/package.xmlParsing /opt/android-sdk-linux/platforms/android-19/package.xmlParsing /opt/android-sdk-linux/platforms/android-21/package.xmlParsing /opt/android-sdk-linux/platforms/android-22/package.xmlParsing /opt/android-sdk-linux/platforms/android-23/package.xmlParsing /opt/android-sdk-linux/platforms/android-24/package.xmlParsing /opt/android-sdk-linux/platforms/android-25/package.xmlParsing /opt/android-sdk-linux/platforms/android-26/package.xmlParsing /opt/android-sdk-linux/platforms/android-27/package.xmlParsing /opt/android-sdk-linux/platforms/android-28/package.xmlParsing /opt/android-sdk-linux/system-images/android-19/default/armeabi-v7a/package.xmlParsing /opt/android-sdk-linux/system-images/android-22/default/armeabi-v7a/package.xmlParsing /opt/android-sdk-linux/system-images/android-24/default/armeabi-v7a/package.xmlParsing /opt/android-sdk-linux/system-images/android-25/google_apis/armeabi-v7a/package.xmlParsing /opt/android-sdk-linux/system-images/android-26/google_apis/x86/package.xmlParsing /opt/android-sdk-linux/system-images/android-28/google_apis/x86/package.xmlParsing /opt/android-sdk-linux/tools/package.xmlAuto-selecting single ABI x86
Do you wish to create a custom hardware profile? [no] Waiting emulator is ready...
Task was aborted
```

**Comment by gigaga** [comment]
Can u copy paste your script step?
Can u reexecute by uncommenting set -x line and remove "> /dev/null" of emulator - avd command?

**Comment by TannerJuby1** [comment]
@gigaga I was actually able to get the tests to run properly by using the built in `Android Build for UI Testing` and `Virtual Device testing for Android` steps. But thanks for your help!

This was the tutorial I used: [https://devcenter.bitrise.io/testing/device-testing-for-android/](url)

**Comment by gigaga** [comment]
On great new! For my part, I didn't want this solution because I didn't want to be dependent of Bitrise infra (just use the engine).

**Comment by muskacirca** [comment]
Hello @gigaga 

I try your workaround but it doesn't seems to work for me. Here's the log without the  "> /dev/null" of emulator -avd command

`Agent version:	1.6.1
Start time:	2019-10-08T08:09:17Z
$ docker "run" "-v" "/var/run/docker.sock:/var/run/docker.sock" "-v" "/bitrise:/bitrise" "-v" "/tmp/bitrise-agent061916546:/.bitrise-config" "-e" "BITRISE_BUILD_NUMBER=125" "-e" "BITRISE_APP_URL=https://app.bitrise.io/app/2b991a7ab303cc8e" "-e" "BITRISE_APP_TITLE=qalyo-android" "-e" "BITRISEIO_GIT_REPOSITORY_OWNER=qalyo-eng" "-e" "BITRISEIO_GIT_REPOSITORY_SLUG=qalyo-android" "-e" "BITRISEIO_STACK_ID=linux-docker-android" "-e" "BITRISE_BUILD_TRIGGER_TIMESTAMP=1570522157" "-e" "BITRISE_BUILD_URL=https://app.bitrise.io/build/9ff4f63708237731" "-e" "BITRISE_BUILD_SLUG=9ff4f63708237731" "-e" "BITRISE_IO=true" "-e" "BITRISE_APP_SLUG=2b991a7ab303cc8e" "-e" "GIT_REPOSITORY_URL=git@bitbucket.org:qalyo-eng/qalyo-android.git" "-e" "BITRISE_GIT_BRANCH=develop" "--privileged" "--network=host" "--name=bitrise-main-container" "bitriseio/android-ndk:pinned" "bitrise" "run" "--config" "/.bitrise-config/bitrise.yml" "--inventory" "/.bitrise-config/bitrise.secrets.yml" "--json-params-base64" "eyJ3b3JrZmxvdyI6InJlcG9ydHMtV0lQIn0="
  ██████╗ ██╗████████╗██████╗ ██╗███████╗███████╗
  ██╔══██╗██║╚══██╔══╝██╔══██╗██║██╔════╝██╔════╝
  ██████╔╝██║   ██║   ██████╔╝██║███████╗█████╗
  ██╔══██╗██║   ██║   ██╔══██╗██║╚════██║██╔══╝
  ██████╔╝██║   ██║   ██║  ██║██║███████║███████╗
  ╚═════╝ ╚═╝   ╚═╝   ╚═╝  ╚═╝╚═╝╚══════╝╚══════╝
  version: 1.34.0
INFO[08:09:18] bitrise runs in Secret Filtering mode 
INFO[08:09:18] bitrise runs in CI mode           
INFO[08:09:18] Running workflow: reports-WIP     
Switching to workflow: reports-WIP
+------------------------------------------------------------------------------+

| (0) activate-ssh-key@4.0.3                                                   |
+---+---------------------------------------------------------------+----------+
+---+---------------------------------------------------------------+----------+
| ✓ | activate-ssh-key@4.0.3                                        | 5.01 sec |
+---+---------------------------------------------------------------+----------+
                                          ▼
INFO[08:09:24] Step uses latest version -- Updating StepLib ... 
INFO[08:09:24] Update StepLib (https://github.com/bitrise-io/bitrise-steplib.git)... 
+------------------------------------------------------------------------------+

| (1) git-clone                                                                |
+------------------------------------------------------------------------------+
| id: git-clone                                                                |
| version: 4.0.17                                                              |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-10-08T08:09:29Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
config:
- RepositoryURL: git@bitbucket.org:qalyo-eng/qalyo-android.git
- CloneIntoDir: /bitrise/src
- Commit: 
- Tag: 
- Branch: develop
- BranchDest: 
- PRID: 0
- PRRepositoryURL: 
- PRMergeBranch: 
- ResetRepository: false
- CloneDepth: 0
- BuildURL: https://app.bitrise.io/build/9ff4f63708237731
- BuildAPIToken: [REDACTED]
- UpdateSubmodules: true
- ManualMerge: true
git "init"
Initialized empty Git repository in /bitrise/src/.git/
git "remote" "add" "origin" "git@bitbucket.org:qalyo-eng/qalyo-android.git"
git "fetch" "origin" "develop"
Warning: Permanently added 'bitbucket.org,18.205.93.1' (RSA) to the list of known hosts.
From bitbucket.org:qalyo-eng/qalyo-android
 * branch            develop    -> FETCH_HEAD
 * [new branch]      develop    -> origin/develop
git "checkout" "develop"
Switched to a new branch 'develop'
Branch develop set up to track remote branch develop from origin.
git "merge" "origin/develop"
Already up-to-date.
git "submodule" "update" "--init" "--recursive"
Exporting git logs
=> GIT_CLONE_COMMIT_MESSAGE_BODY
   value: 
=> GIT_CLONE_COMMIT_AUTHOR_NAME
   value: Vincent Truchot
=> GIT_CLONE_COMMIT_AUTHOR_EMAIL
   value: vincent.truchot@qalyo.com
=> GIT_CLONE_COMMIT_COMMITER_NAME
   value: Vincent Truchot
=> GIT_CLONE_COMMIT_COMMITER_EMAIL
   value: vincent.truchot@qalyo.com
=> GIT_CLONE_COMMIT_HASH
   value: 3fc18f46128c0657c72d1efce038097316b39ba6
=> GIT_CLONE_COMMIT_MESSAGE_SUBJECT
   value: QPB-641 : Renamed connected objects to data sources
=> GIT_CLONE_COMMIT_COUNT
   value: 6443
Success
|                                                                              |
+---+---------------------------------------------------------------+----------+
| ✓ | git-clone                                                     | 24.69 sec|
+---+---------------------------------------------------------------+----------+
                                          ▼
+------------------------------------------------------------------------------+
| (2) cache-pull@2.1.0                                                         |
+------------------------------------------------------------------------------+
| id: cache-pull                                                               |
| version: 2.1.0                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2019-10-08T08:09:50Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
INFO[08:09:50] Start installing (tar) with apt-get          
INFO[08:09:50]  * [OK] Step dependency (tar) installed, available. 
Config:
- CacheAPIURL: [REDACTED]
- DebugMode: false
- StackID: 
Downloading remote cache archive
Failed to get cache download url: build cache not found: probably cache not initialised yet (first cache push initialises the cache), nothing to worry about ;)
WARN[08:09:53] Step (cache-pull@2.1.0) failed, but was marked as skippable 
|                                                                              |
+---+---------------------------------------------------------------+----------+
| ! | cache-pull@2.1.0 (exit code: 1)                               | 4.70 sec |
+---+---------------------------------------------------------------+----------+
| Update available: 2.1.0 -> 2.1.1                                             |
| Issue tracker: https://github.com/bitrise-steplib/steps-cache-pull/issues    |
| Source: https://github.com/bitrise-steplib/steps-cache-pull.git              |
+---+---------------------------------------------------------------+----------+
                                          ▼
+------------------------------------------------------------------------------+

| (3) install-missing-android-tools@2.3.7                                      |
+---+---------------------------------------------------------------+----------+
+---+---------------------------------------------------------------+----------+
| ✓ | install-missing-android-tools@2.3.7                           | 1.7 min  |
+---+---------------------------------------------------------------+----------+
                                          ▼
+------------------------------------------------------------------------------+

| (4) script@1.1.5                                                             |
+------------------------------------------------------------------------------+
| id: script                                                                   |
| version: 1.1.5                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: bash                                                                |
| time: 2019-10-08T08:11:38Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
+ cd /opt/android-sdk-linux/emulator
+ avdmanager create avd -n Nexus_5X_API_26 -k 'system-images;android-26;google_apis;x86' --force
+ echo no
Loading local repository...                                                     
[=========                              ] 25% Loading local repository...       
[=========                              ] 25% Fetch remote repository...        
[=========                              ] 25% Fetch remote repository...        
[=========                              ] 25% Fetch remote repository...        
[=======================================] 100% Fetch remote repository...       
Auto-selecting single ABI x86
Do you wish to create a custom hardware profile? [no] + emulator -avd Nexus_5X_API_26 -no-window -no-audio -debug-init
statvfs('/root/.android/avd/Nexus_5X_API_26.avd/snapshots/default_boot/ram.img') failed: No such file or directory
emulator: ERROR: AdbHostServer.cpp:93: Unable to connect to adb daemon on port: 5037
emulator: ERROR: AdbHostServer.cpp:93: Unable to connect to adb daemon on port: 5037
emulator: autoconfig: -skin HVGA
emulator: autoconfig: -skindir (null)
emulator: autoconfig: -kernel /opt/android-sdk-linux/system-images/android-26/google_apis/x86//kernel-ranchu
emulator: Target arch = 'x86'
emulator: Auto-config: -qemu -cpu qemu32
emulator: Auto-detect: Kernel image requires new device naming scheme.
emulator: Auto-detect: Kernel does not support YAFFS2 partitions.
emulator: autoconfig: -ramdisk /opt/android-sdk-linux/system-images/android-26/google_apis/x86//ramdisk.img
emulator: Using initial system image: /opt/android-sdk-linux/system-images/android-26/google_apis/x86//system.img
emulator: Using initial vendor image: /opt/android-sdk-linux/system-images/android-26/google_apis/x86//vendor.img
emulator: autoconfig: -initdata /root/.android/avd/Nexus_5X_API_26.avd/userdata.img
emulator: autoconfig: -cache /root/.android/avd/Nexus_5X_API_26.avd/cache.img
emulator: Increasing RAM size to 1536MB
emulator: VM heap size 0MB is below hardware specified minimum of 384MB,setting it to that value
emulator: System image is read only
emulator: Found 1 DNS servers: 169.254.169.254
emulator: skin name 'HVGA' aliased to '320x480'
emulator: found magic skin width=320 height=480 bpp=16
emulator: registered 'boot-properties' qemud service
emulator: Adding boot property: 'qemu.cmdline' = 'androidboot.logcat=v:* androidboot.consolepipe=qemu_pipe,pipe:logcat'
emulator: Adding boot property: 'qemu.logcat' = 'start'
emulator: Creating ext4 userdata partition: /root/.android/avd/Nexus_5X_API_26.avd/data
emulator: cannot read adb key file: /root/.android/adbkey.pub
emulator: trying again by copying from home dir
emulator: path_copy_file: source file is un-readable: /root/.android/adbkey.pub
emulator: cannot read adb key file (failed): /root/.android/adbkey.pub
emulator: cannot read adb key file: /root/.android/adbkey
emulator: trying again by copying from home dir
emulator: path_copy_file: source file is un-readable: /root/.android/adbkey
emulator: cannot read adb key file (failed): /root/.android/adbkey
emulator: generate_key(/root/.android/adbkey)...
emulator: Fall back to adbkey /root/.android/adbkey successfully
Creating filesystem with parameters:
    Size: 838860800
    Block size: 4096
    Blocks per group: 32768
    Inodes per group: 7328
    Inode size: 256
    Journal blocks: 3200
    Label: 
    Blocks: 204800
    Block groups: 7
    Reserved block group size: 55
Created filesystem with 74/51296 inodes and 13058/204800 blocks
emulator: Creating empty ext4 cache partition: /root/.android/avd/Nexus_5X_API_26.avd/cache.img
Creating filesystem with parameters:
    Size: 69206016
    Block size: 4096
    Blocks per group: 32768
    Inodes per group: 4224
    Inode size: 256
    Journal blocks: 1024
    Label: 
    Blocks: 16896
    Block groups: 1
    Reserved block group size: 7
Created filesystem with 11/4224 inodes and 1302/16896 blocks
emulator: CPU Acceleration: working
emulator: CPU Acceleration status: KVM (version 12) is installed and usable.
emulator: GPU emulation enabled using 'swiftshader_indirect' mode
emulator: Initializing hardware OpenGLES emulation support
emulator: Found 1 DNS servers: 169.254.169.254
Content of hardware configuration file:
  hw.cpu.arch = x86
  hw.cpu.model = qemu32
  hw.cpu.ncore = 2
  hw.ramSize = 1536
  hw.screen = multi-touch
  hw.mainKeys = true
  hw.trackBall = true
  hw.keyboard = false
  hw.keyboard.lid = false
  hw.keyboard.charmap = qwerty2
  hw.dPad = true
  hw.rotaryInput = false
  hw.gsmModem = true
  hw.gps = true
  hw.battery = true
  hw.accelerometer = true
  hw.gyroscope = true
  hw.audioInput = true
  hw.audioOutput = true
  hw.sdCard = true
  disk.cachePartition = true
  disk.cachePartition.path = /root/.android/avd/Nexus_5X_API_26.avd/cache.img
  disk.cachePartition.size = 66m
  test.quitAfterBootTimeOut = -1
  test.delayAdbTillBootComplete = 0
  test.monitorAdb = 0
  hw.lcd.width = 320
  hw.lcd.height = 480
  hw.lcd.depth = 16
  hw.lcd.density = 160
  hw.lcd.backlight = true
  hw.displayRegion.0.1.xOffset = -1
  hw.displayRegion.0.1.yOffset = -1
  hw.displayRegion.0.1.width = 0
  hw.displayRegion.0.1.height = 0
  hw.fold.adjust = false
  hw.display1.width = 0
  hw.display1.height = 0
  hw.display1.density = 0
  hw.display1.xOffset = -1
  hw.display1.yOffset = -1
  hw.display1.flag = 0
  hw.display2.width = 0
  hw.display2.height = 0
  hw.display2.density = 0
  hw.display2.xOffset = -1
  hw.display2.yOffset = -1
  hw.display2.flag = 0
  hw.display3.width = 0
  hw.display3.height = 0
  hw.display3.density = 0
  hw.display3.xOffset = -1
  hw.display3.yOffset = -1
  hw.display3.flag = 0
  hw.gpu.enabled = true
  hw.gpu.mode = swiftshader_indirect
  hw.initialOrientation = portrait
  hw.camera.back = emulated
  hw.camera.front = none
  vm.heapSize = 384
  hw.sensors.light = true
  hw.sensors.pressure = true
  hw.sensors.humidity = true
  hw.sensors.proximity = true
  hw.sensors.magnetic_field = true
  hw.sensors.magnetic_field_uncalibrated = true
  hw.sensors.gyroscope_uncalibrated = true
  hw.sensors.orientation = true
  hw.sensors.temperature = true
  hw.useext4 = true
  hw.arc = false
  hw.arc.autologin = false
  kernel.path = /opt/android-sdk-linux/system-images/android-26/google_apis/x86//kernel-ranchu
  kernel.newDeviceNaming = yes
  kernel.supportsYaffs2 = no
  disk.ramdisk.path = /opt/android-sdk-linux/system-images/android-26/google_apis/x86//ramdisk.img
  disk.systemPartition.initPath = /opt/android-sdk-linux/system-images/android-26/google_apis/x86//system.img
  disk.systemPartition.size = 2562m
  disk.vendorPartition.initPath = /opt/android-sdk-linux/system-images/android-26/google_apis/x86//vendor.img
  disk.vendorPartition.size = 800m
  disk.dataPartition.path = /root/.android/avd/Nexus_5X_API_26.avd/userdata-qemu.img
  disk.dataPartition.initPath = /root/.android/avd/Nexus_5X_API_26.avd/userdata.img
  disk.dataPartition.size = 800m
  disk.encryptionKeyPartition.path = /root/.android/avd/Nexus_5X_API_26.avd/encryptionkey.img
  PlayStore.enabled = false
  avd.name = Nexus_5X_API_26
  avd.id = Nexus_5X_API_26
  fastboot.forceColdBoot = false
.
QEMU options list:
emulator: argv[00] = "/opt/android-sdk-linux/emulator/./qemu/linux-x86_64/qemu-system-i386"
emulator: argv[01] = "-dns-server"
emulator: argv[02] = "169.254.169.254"
emulator: argv[03] = "-no-audio"
emulator: argv[04] = "-mem-path"
emulator: argv[05] = "/root/.android/avd/Nexus_5X_API_26.avd/snapshots/default_boot/ram.img"
emulator: argv[06] = "-mem-file-shared"
emulator: argv[07] = "-serial"
emulator: argv[08] = "null"
emulator: argv[09] = "-device"
emulator: argv[10] = "goldfish_pstore,addr=0xff018000,size=0x10000,file=/root/.android/avd/Nexus_5X_API_26.avd/data/misc/pstore/pstore.bin"
emulator: argv[11] = "-cpu"
emulator: argv[12] = "android32"
emulator: argv[13] = "-enable-kvm"
emulator: argv[14] = "-smp"
emulator: argv[15] = "cores=2"
emulator: argv[16] = "-m"
emulator: argv[17] = "1536"
emulator: argv[18] = "-lcd-density"
emulator: argv[19] = "160"
emulator: argv[20] = "-object"
emulator: argv[21] = "iothread,id=disk-iothread"
emulator: argv[22] = "-nodefaults"
emulator: argv[23] = "-kernel"
emulator: argv[24] = "/opt/android-sdk-linux/system-images/android-26/google_apis/x86//kernel-ranchu"
emulator: argv[25] = "-initrd"
emulator: argv[26] = "/opt/android-sdk-linux/system-images/android-26/google_apis/x86//ramdisk.img"
emulator: argv[27] = "-drive"
emulator: argv[28] = "if=none,index=0,id=system,file=/opt/android-sdk-linux/system-images/android-26/google_apis/x86//system.img,read-only"
emulator: argv[29] = "-device"
emulator: argv[30] = "virtio-blk-pci,drive=system,iothread=disk-iothread,modern-pio-notify"
emulator: argv[31] = "-drive"
emulator: argv[32] = "if=none,index=1,id=cache,file=/root/.android/avd/Nexus_5X_API_26.avd/cache.img.qcow2,overlap-check=none,cache=unsafe,l2-cache-size=1048576"
emulator: argv[33] = "-device"
emulator: argv[34] = "virtio-blk-pci,drive=cache,iothread=disk-iothread,modern-pio-notify"
emulator: argv[35] = "-drive"
emulator: argv[36] = "if=none,index=2,id=userdata,file=/root/.android/avd/Nexus_5X_API_26.avd/userdata-qemu.img.qcow2,overlap-check=none,cache=unsafe,l2-cache-size=1048576"
emulator: argv[37] = "-device"
emulator: argv[38] = "virtio-blk-pci,drive=userdata,iothread=disk-iothread,modern-pio-notify"
emulator: argv[39] = "-drive"
emulator: argv[40] = "if=none,index=3,id=encrypt,file=/root/.android/avd/Nexus_5X_API_26.avd/encryptionkey.img.qcow2,overlap-check=none,cache=unsafe,l2-cache-size=1048576"
emulator: argv[41] = "-device"
emulator: argv[42] = "virtio-blk-pci,drive=encrypt,iothread=disk-iothread,modern-pio-notify"
emulator: argv[43] = "-drive"
emulator: argv[44] = "if=none,index=4,id=vendor,file=/opt/android-sdk-linux/system-images/android-26/google_apis/x86//vendor.img,read-only"
emulator: argv[45] = "-device"
emulator: argv[46] = "virtio-blk-pci,drive=vendor,iothread=disk-iothread,modern-pio-notify"
emulator: argv[47] = "-netdev"
emulator: argv[48] = "user,id=mynet"
emulator: argv[49] = "-device"
emulator: argv[50] = "virtio-net-pci,netdev=mynet"
emulator: argv[51] = "-device"
emulator: argv[52] = "virtio-rng-pci"
emulator: argv[53] = "-show-cursor"
emulator: argv[54] = "-nographic"
emulator: argv[55] = "-monitor"
emulator: argv[56] = "none"
emulator: argv[57] = "-L"
emulator: argv[58] = "/opt/android-sdk-linux/emulator/lib/pc-bios"
emulator: argv[59] = "-soundhw"
emulator: argv[60] = "hda"
emulator: argv[61] = "-vga"
emulator: argv[62] = "none"
emulator: argv[63] = "-append"
emulator: argv[64] = "qemu=1 no_timer_check androidboot.hardware=ranchu androidboot.serialno=EMULATOR29X2X2X0 clocksource=pit no-kvmclock android.qemud=1 console=0 android.checkjni=1 qemu.gles=1 qemu.settings.system.screen_off_timeout=2147483647 qemu.encrypt=1 qemu.opengles.version=131072 cma=260M@0-4G qemu.wifi=1 mac80211_hwsim.channels=2 androidboot.android_dt_dir=/sys/bus/platform/devices/ANDR0001:00/properties/android/ loop.max_part=7 ramoops.mem_address=0xff018000 ramoops.mem_size=0x10000 memmap=0x10000$0xff018000 qemu.dalvik.vm.heapsize=384m"
emulator: argv[65] = "-android-hw"
emulator: argv[66] = "/root/.android/avd/Nexus_5X_API_26.avd/hardware-qemu.ini"
Concatenated QEMU options:
 /opt/android-sdk-linux/emulator/./qemu/linux-x86_64/qemu-system-i386 -dns-server 169.254.169.254 -no-audio -mem-path /root/.android/avd/Nexus_5X_API_26.avd/snapshots/default_boot/ram.img -mem-file-shared -serial null -device goldfish_pstore,addr=0xff018000,size=0x10000,file=/root/.android/avd/Nexus_5X_API_26.avd/data/misc/pstore/pstore.bin -cpu android32 -enable-kvm -smp cores=2 -m 1536 -lcd-density 160 -object iothread,id=disk-iothread -nodefaults -kernel /opt/android-sdk-linux/system-images/android-26/google_apis/x86//kernel-ranchu -initrd /opt/android-sdk-linux/system-images/android-26/google_apis/x86//ramdisk.img -drive if=none,index=0,id=system,file=/opt/android-sdk-linux/system-images/android-26/google_apis/x86//system.img,read-only -device virtio-blk-pci,drive=system,iothread=disk-iothread,modern-pio-notify -drive if=none,index=1,id=cache,file=/root/.android/avd/Nexus_5X_API_26.avd/cache.img.qcow2,overlap-check=none,cache=unsafe,l2-cache-size=1048576 -device virtio-blk-pci,drive=cache,iothread=disk-iothread,modern-pio-notify -drive if=none,index=2,id=userdata,file=/root/.android/avd/Nexus_5X_API_26.avd/userdata-qemu.img.qcow2,overlap-check=none,cache=unsafe,l2-cache-size=1048576 -device virtio-blk-pci,drive=userdata,iothread=disk-iothread,modern-pio-notify -drive if=none,index=3,id=encrypt,file=/root/.android/avd/Nexus_5X_API_26.avd/encryptionkey.img.qcow2,overlap-check=none,cache=unsafe,l2-cache-size=1048576 -device virtio-blk-pci,drive=encrypt,iothread=disk-iothread,modern-pio-notify -drive if=none,index=4,id=vendor,file=/opt/android-sdk-linux/system-images/android-26/google_apis/x86//vendor.img,read-only -device virtio-blk-pci,drive=vendor,iothread=disk-iothread,modern-pio-notify -netdev user,id=mynet -device virtio-net-pci,netdev=mynet -device virtio-rng-pci -show-cursor -nographic -monitor none -L /opt/android-sdk-linux/emulator/lib/pc-bios -soundhw hda -vga none -append 'qemu=1 no_timer_check androidboot.hardware=ranchu androidboot.serialno=EMULATOR29X2X2X0 clocksource=pit no-kvmclock android.qemud=1 console=0 android.checkjni=1 qemu.gles=1 qemu.settings.system.screen_off_timeout=2147483647 qemu.encrypt=1 qemu.opengles.version=131072 cma=260M@0-4G qemu.wifi=1 mac80211_hwsim.channels=2 androidboot.android_dt_dir=/sys/bus/platform/devices/ANDR0001:00/properties/android/ loop.max_part=7 ramoops.mem_address=0xff018000 ramoops.mem_size=0x10000 memmap=0x10000$0xff018000 qemu.dalvik.vm.heapsize=384m' -android-hw /root/.android/avd/Nexus_5X_API_26.avd/hardware-qemu.ini
emulator: Android qemu version 29.2.2.0 (build_id 5904653) (CL:N/A)
emulator: Starting QEMU main loop
emulator: Adding boot property: 'ro.opengles.version' = '131072'
emulator: Adding boot property: 'qemu.sf.fake_camera' = 'back'
emulator: Adding boot property: 'dalvik.vm.heapsize' = '384m'
emulator: Adding boot property: 'qemu.hw.mainkeys' = '1'
emulator: Adding boot property: 'qemu.sf.lcd_density' = '160'
emulator: goldfish_events.have-dpad: true
emulator: goldfish_events.have-trackball: true
emulator: goldfish_events.have-camera: true
emulator: goldfish_events.have-keyboard: false
emulator: goldfish_events.have-lidswitch: false
emulator: goldfish_events.have-tabletmode: false
emulator: goldfish_events.have-touch: false
emulator: goldfish_events.have-multitouch: true
emulator: control console listening on port 5554, ADB on port 5555
Not using any http proxy
emulator: Adding boot property: 'qemu.timezone' = 'Etc/UTC'
emulator: android_hw_fingerprint_init: fingerprint qemud listen service initialized
emulator: No acpi ini file provided, using default
emulator: No acpi ini file provided, using default
Your emulator is out of date, please update by launching Android Studio:
 - Start Android Studio
 - Select menu "Tools > Android > SDK Manager"
 - Click "SDK Tools" tab
 - Check "Android Emulator" checkbox
 - Click "OK"
emulator: onGuestSendCommand: [0x84d7e00] Adb connected, start proxing data
emulator: _hwFingerprint_connect: connect finger print listen is called
emulator: got message from guest system fingerprint HAL
`

then it timeout.

Do you see anythins i'm doing wrong ?

Thank you

**Comment by gigaga** [comment]
I continue to use this workaround, so it still works.
Can you provide your script used?

**Comment by muskacirca** [comment]
Thank your for your quick answer.

Actually I managed to make your script works (unchanged). 
I've placed it as first step and I changed my stack to LTS !

Boom ! Now my tests are failing ;)

**Comment by gigaga** [comment]
:)

**Comment by ofalvai** [comment]
For anyone stumbling on this old thread: [version 1.0.7](https://github.com/bitrise-steplib/steps-wait-for-android-emulator/releases/tag/1.0.7) of this step includes improvements to the waiting logic, similar to the shell script posted earlier in this thread.

If you still experience issues with the step's stability, please open a new issue.
